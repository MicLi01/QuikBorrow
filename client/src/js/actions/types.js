export const GET_ERRORS = "GET_ERRORS";
export const USER_LOADING = "USER_LOADING";
export const SET_CURRENT_USER = "SET_CURRENT_USER";
export const SEARCH = "SEARCH";
export const GET_USER_PROFILE = "GET_USER_PROFILE";

// Products
export const FETCH_PRODUCTS = "fetch_products";
export const FETCH_PRODUCT = "fetch_product";
export const BORROW_PRODUCT_FETCH = 'BORROW_PRODUCT_FETCH';
export const BORROW_PRODUCT_FINISHED = 'BORROW_PRODUCT_FINISHED';

