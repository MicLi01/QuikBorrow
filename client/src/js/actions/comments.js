import axios from "axios";
import { CREATE_COMMENT, GET_COMMENTS } from './types';

/**
 * @typedef {import('redux').Dispatch} DispatchFunction
 */

/**
 * Create a comment
 * 
 * @param {Object} comment - The Comment object to create
 * @returns {DispatchFunction}
 */
export const createComment = (comment) => dispatch => {
  return axios
    .post("/api/comments", comment)
    .then(res => {
    	console.log(res);
    	console.log(comment);
    })
    .catch(err => {
    	console.log(err);
  		dispatch({
  			type: CREATE_COMMENT.ERROR,
  			payload: err.response.data
  		})
    });
};

/**
 *  Get comments associated with product
 *  @param {String} id - object id of product
 *  @returns {DispatchFunction} dispatch - used to dispatch actions
 */
export const getComments = (id) => dispatch => {
  return axios
    .get(`/api/comments/${id}`)
    .then(res => {
      dispatch({
        type: GET_COMMENTS.FINISHED,
        payload: res.data
      });
    })
    .catch(err => {
      dispatch({
        type: GET_COMMENTS.ERROR,
        payload: err.response.data
      })
    });
};