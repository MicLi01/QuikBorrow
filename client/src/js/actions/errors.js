import { CLEAR_ALL_ERRORS } from './types'

/**
 * Clears all error data in error state for all error types
 */
 export const clearAllErrors = () => {
     return { type: CLEAR_ALL_ERRORS }
 }
