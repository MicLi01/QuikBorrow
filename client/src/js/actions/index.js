import axios from "axios";
import setAuthToken from "../utils/setAuthToken";
import jwt_decode from "jwt-decode";
import { FETCH_PRODUCTS, FETCH_PRODUCT, SET_CURRENT_USER, USER_LOADING, SEARCH_PRODUCTS, GET_USER, REGISTER_USER, LOGIN_USER } from "./types";
export * from './products';

// Fetch all products
export const fetchProducts = () => {
  return async function(dispatch) {
    const res = await axios.get("/api/items");
    dispatch({ type: FETCH_PRODUCTS, payload: res.data });
  };
};

// Get one product
export const fetchProduct = itemId => {
  return async function(dispatch) {
    const res = await axios.get(`/api/items/${itemId}`);
    dispatch({ type: FETCH_PRODUCT, payload: res.data });
  };
};

/**
 *  Search by query for items in database by calling {@link Action} to backend
 *  @param {String} query - query string to search
 *  @returns {Function} dispatch - used to dispatch actions
 */
export const searchProducts = query => 
  /**
   *  @param {Function} dispatch - used to dispatch actions
   *  @returns {Promise<Object>} - response from an {@link Action}
   */
  dispatch => {
    return axios
        .get("/api/items/search", { params: { param: query} })
        .then(res => {
          dispatch({ type: SEARCH_PRODUCTS.FETCH, payload: res.data });
        })
        .catch(err => {
          dispatch({
            type: SEARCH_PRODUCTS.ERROR,
            payload: err.response.data
          })
        });
  };

// Register User
export const registerUser = (userData, history) => dispatch => {
  return axios
    .post("/api/users/register", userData)
    .then(res => history.push("/login")) // re-direct to login on successful register
    .catch(err =>
      dispatch({
        type: REGISTER_USER.ERROR,
        payload: err.response.data
      })
    );
};

// Login - get user token
export const loginUser = userData => dispatch => {
  return axios
    .post("/api/users/login", userData)
    .then(res => {
      // Save to localStorage
      // Set token to localStorage
      const { token } = res.data;
      localStorage.setItem("jwtToken", token);
      // Set token to Auth header
      setAuthToken(token);
      // Decode token to get user data
      const decoded = jwt_decode(token);
      // Set current user
      dispatch(setCurrentUser(decoded));
    })
    .catch(err =>
      dispatch({
        type: LOGIN_USER.ERROR,
        payload: err.response.data
      })
    );
};

// Set logged in user
export const setCurrentUser = decoded => {
  return {
    type: SET_CURRENT_USER,
    payload: decoded
  };
};

// User loading
export const setUserLoading = () => {
  return {
    type: USER_LOADING
  };
};

// Log user out
export const logoutUser = () => dispatch => {
  // Remove token from local storage
  localStorage.removeItem("jwtToken");
  // Remove auth header for future requests
  setAuthToken(false);
  // Set current user to empty object {} which will set isAuthenticated to false
  dispatch(setCurrentUser({}));
};

export const getUserProfile = userId => {
  return async function(dispatch) {
    try {
    const res = await axios.get(`/api/users/${userId}`);
    dispatch({ type: GET_USER.FINISHED, payload: res.data });
  }
  catch(error) {
    console.log(error)
    dispatch({
      type: GET_USER.ERROR, payload: error
      });
    }
  }
};
