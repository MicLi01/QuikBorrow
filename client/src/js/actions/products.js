import axios from "axios";
import {
    FETCH_PRODUCTS,
    FETCH_PRODUCT,
    SEARCH_PRODUCTS,
    REQUEST_BORROW_PRODUCT,
    FETCH_TRANSACTIONS
} from './types'

/**
 * @typedef {import('redux').Dispatch} DispatchFunction
 */

/**
 * Borrow an item via fetch
 * 
 * @param {Object} item - The Item object to borrow
 * @param {string} msg - The message for the lender
 * @returns {DispatchFunction}
 */
export const requestBorrowProductFetch = (item, msg) => async (dispatch, getState) => {
    const state = getState()
    const body = {
        borrowerId: state.auth.user.id,
        lenderId: item.user,
        itemId: item._id,
        msg
    }
    return axios.post(`/api/transactions`, body)
        .then(({ data }) => {
            dispatch({ type: REQUEST_BORROW_PRODUCT.FINISHED, payload: data })
        })
        .catch(err => {
            dispatch({ type: REQUEST_BORROW_PRODUCT.ERROR, payload: err.response.data })
        })
}

/**
 * Fetch all products
 * @returns {DispatchFunction}
 */
export const fetchProducts = () => {
  return async function(dispatch) {
    const res = await axios.get("/api/items");
    dispatch({ type: FETCH_PRODUCTS, payload: res.data });
  };
};

/**
 * Get one product
 * 
 * @param {Object} itemId - The id of the item to retrieve
 * @returns {DispatchFunction}
 */
export const fetchProduct = itemId => {
  return async function(dispatch) {
    const res = await axios.get(`/api/items/${itemId}`);
    dispatch({ type: FETCH_PRODUCT, payload: res.data });
  };
};

/**
 *  Search by query for items in database by calling {@link Action} to backend
 *  @param {String} query - query string to search
 *  @returns {Function} dispatch - used to dispatch actions
 */
export const searchProducts = query => 
  /**
   *  @param {Function} dispatch - used to dispatch actions
   *  @returns {Promise<Object>} - response from an {@link Action}
   */
  dispatch => {
    return axios
        .get("/api/items/search", { params: { param: query} })
        .then(res => {
          dispatch({ type: SEARCH_PRODUCTS.FINISHED, payload: res.data });
        })
        .catch(err => {
          dispatch({
            type: SEARCH_PRODUCTS.ERROR,
            payload: err.response.data
          })
        });
  };

  /**
 * Get all current user transactions
 * 
 * @param {Object} id - The user id
 * @returns {Function} dispatch - used to dispatch actions
 */

 export const fetchTransactions = (id) => 
  /**
   *  @param {Function} dispatch - used to dispatch actions
   *  @returns {Promise<Object>} - response from an {@link Action}
   */
  dispatch => {
    return axios.post('/api/transactions/getTransactions', {userId : id})
        .then(res => dispatch({ type: FETCH_TRANSACTIONS.FINISHED, payload: res.data }))
        .catch(err => dispatch({ type: FETCH_TRANSACTIONS.ERROR, payload: err.response.data }))
 }

