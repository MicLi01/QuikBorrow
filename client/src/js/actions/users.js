import axios from "axios";
import setAuthToken from "../utils/setAuthToken";
import jwt_decode from "jwt-decode";
import { SET_CURRENT_USER, USER_LOADING, REGISTER_USER, LOGIN_USER, GET_USER } from "./types";

/**
 * @typedef {import('redux').Dispatch} DispatchFunction
 */

/**
 *	Register user 
 *	@param {Object} userData - User credentials to create an account
 *	@param {Object} history - used to navigate to different pages
 *	@returns {DispatchFunction}
 */
export const registerUser = (userData, history) => dispatch => {
  return axios
    .post("/api/users/register", userData)
    .then(res => history.push("/login")) // re-direct to login on successful register
    .catch(err =>
      dispatch({
        type: REGISTER_USER.ERROR,
        payload: err.response.data
      })
    );
};

/**
 *	Login user 
 *	@param {Object} userData - User credentials to login
 *	@returns {DispatchFunction}
 */
export const loginUser = userData => dispatch => {
  return axios
    .post("/api/users/login", userData)
    .then(res => {
      const { token } = res.data;
      localStorage.setItem("jwtToken", token);
      // Set token to Auth header
      setAuthToken(token);
      // Decode token to get user data
      const decoded = jwt_decode(token);
      // Set current user
      dispatch(setCurrentUser(decoded));
    })
    .catch(err =>
      dispatch({
        type: LOGIN_USER.ERROR,
        payload: err.response.data
      })
    );
};

/**
 *	Set current user
 *	@param {Object} decoded - decoded jwtToken containing user info
 *	@returns {Object}
 */
export const setCurrentUser = decoded => {
  return {
    type: SET_CURRENT_USER,
    payload: decoded
  };
};

/**
 *	Set user loading
 *	@returns {Object}
 */
export const setUserLoading = () => {
  return {
    type: USER_LOADING
  };
};

/**
 *	Logout user by removing auth token and saved info in storage
 *	@returns {DispatchFunction}
 */
export const logoutUser = () => dispatch => {
  // Remove token from local storage
  localStorage.removeItem("jwtToken");
  // Remove auth header for future requests
  setAuthToken(false);
  // Set current user to empty object {} which will set isAuthenticated to false
  dispatch(setCurrentUser({}));
};

/**
 *	Get user profile
 *	@param {Object} userId - get profile based on userId
 *	@returns {DispatchFunction}
 */
export const getUserProfile = userId => {
  return async function(dispatch) {
    try {
    const res = await axios.get(`/api/users/${userId}`);
    dispatch({ type: GET_USER.FINISHED, payload: res.data });
  }
  catch(error) {
    console.log(error)
    dispatch({
      type: GET_USER.ERROR, payload: error
      });
    }
  }
};