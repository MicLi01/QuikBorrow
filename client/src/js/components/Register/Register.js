import React, { Component } from "react";
import { Link, withRouter } from "react-router-dom";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { registerUser } from "../../actions/index";
import classnames from "classnames";
import styled from 'styled-components';
import { Button, TextField } from '@material-ui/core';
import quikLogo from '../../components/quikLogo.png';

const LoginForm = styled.div`
	margin: 0 auto;
    max-width: 320px;
	outline: 1px solid #CCC;
	box-shadow: 5px #CCC;
    background-color: white;
	padding:1.5em;
	box-shadow: inset 0 1.5px 2px rgba(190, 190, 190, .4), 0 0 0 5px #f5f7f8;
`
const Center = styled.div`
	text-align: center;
`
  
export class Register extends Component {
	constructor() {
		super();
		this.state = {
			name: "",
			email: "",
			password: "",
			password2: "",
			errors: {}
		};
	}

	componentDidMount() {
		// If logged in and user navigates to Register page, should redirect them to dashboard
		if (this.props.auth.isAuthenticated) {
		  this.props.history.push("/dashboard");
		}
	}

	componentWillReceiveProps(nextProps) {
    	if (nextProps.errors) {
      		this.setState({
        		errors: nextProps.errors
      		});
    	}
  	}
	onChange = e => {
		this.setState({ [e.target.id]: e.target.value });
	};
	onSubmit = e => {
		e.preventDefault();
		const newUser = {
			name: this.state.name,
			email: this.state.email,
			password: this.state.password,
			password2: this.state.password2
		};
		this.props.registerUser(newUser, this.props.history); 
	};
	getErrors = e => {
		const { errors } = this.state;
		if(errors.errors === undefined || errors.errors.find(x => x.param === e) === undefined)
		{
			return "";
		}
		else
		{
			return errors.errors.find(x => x.param === e).msg;
		}
	};
	render() {
		
		return (
			<div>
				<Center>
					<img src={quikLogo} width="275" height= "180" alt = "logo"/>
				</Center>
				<LoginForm>
					<h4>New User? Register below:</h4>
					<p>Already have an account? <Link to="/login">Log in</Link></p>
				<form noValidate onSubmit={this.onSubmit}>
					<TextField
						onChange={this.onChange}
						value={this.state.name}
						error={this.getErrors('name')!==""}
						helperText={this.getErrors('name')}
						id="name"
						label="Name"
						type="text"
						fullWidth
						className={classnames("", {
							invalid: this.getErrors('name')
		                })}
					/>
					<br/>
					<TextField
						onChange={this.onChange}
						value={this.state.email}
						error={this.getErrors('email')!==""}
						helperText={this.getErrors('email')}
						id="email"
						type="email"
						label="Email"
						fullWidth
						className={classnames("", {
                    		invalid: this.getErrors('email')
                  		})}
					/>
					<br/>
					<TextField
						onChange={this.onChange}
						value={this.state.password}
						error={this.getErrors('password')!==""}
						helperText={this.getErrors('password')}
						id="password"
						type="password"
						label="Password"
						fullWidth
						className={classnames("", {
                    		invalid: this.getErrors('password')
                  		})}
					/>
					<br/>
					<TextField
						onChange={this.onChange}
						value={this.state.password2}
						error={this.getErrors('password')!==""}
						helperText={this.getErrors('password')}
						id="password2"
						type="password"
						label="Confirm Password"
						fullWidth
						className={classnames("", {
                    		invalid: this.getErrors('password')
                  		})}
					/>
					<br/>
					<br/>
					<div>
					<Button 
						variant="contained" 
						color="primary" 
						type="submit"
						fullWidth
					>
        			Register
      				</Button>
					</div>
				</form>
				</LoginForm>
			</div>
		);
	}
}

Register.propTypes = {
  registerUser: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
  errors: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth,
  errors: state.errors
});

export default connect(
  mapStateToProps,
  { registerUser }
)(withRouter(Register));