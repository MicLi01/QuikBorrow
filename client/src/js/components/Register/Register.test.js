import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Register } from './Register';
import { registerUser } from "../../actions/index";

Enzyme.configure({ adapter: new Adapter() });
jest.mock('../../actions/index');

describe('Register', () => {
	const params = {
		registerUser: registerUser,
		auth: {
			isAuthenticated: false,
			user: {},
			loading: false
		},
		errors: {}
	}

	test('renders', () => {
		const component = renderer.create(
			<MemoryRouter>
				<Register registerUser = {params.registerUser} auth = {params.auth} errors = {params.errors} />
			</MemoryRouter>
		).toJSON();
		
		expect(component).toMatchSnapshot();
	});
	
	test('username entry', () => {
		let wrapper = shallow(<Register registerUser = {params.registerUser} auth = {params.auth} errors = {params.errors} />);
		wrapper.find('InputText[type="text"]').simulate('change', {target: {id: 'name', value: 'test'}});
		expect(wrapper.state('name')).toEqual('test');
	});

	test('email entry', () => {
		let wrapper = shallow(<Register registerUser = {params.registerUser} auth = {params.auth} errors = {params.errors} />);
		wrapper.find('InputText[type="text"]').simulate('change', {target: {id: 'email', value: 'test@test.com'}});
		expect(wrapper.state('email')).toEqual('test@test.com');
	});

	test('password entry', () => {
		let wrapper = shallow(<Register registerUser = {params.registerUser} auth = {params.auth} errors = {params.errors} />);
		wrapper.find('InputText[type="text"]').simulate('change', {target: {id: 'password', value: 'password'}});
		expect(wrapper.state('password')).toEqual('password');
	});

	test('password confirmation entry', () => {
		let wrapper = shallow(<Register registerUser = {params.registerUser} auth = {params.auth} errors = {params.errors} />);
		wrapper.find('InputText[type="text"]').simulate('change', {target: {id: 'password2', value: 'password'}});
		expect(wrapper.state('password2')).toEqual('password');
	});

	test('successful registration', () => {
		let wrapper = shallow(<Register registerUser = {params.registerUser} auth = {params.auth} errors = {params.errors} />);
		wrapper.find('InputText[type="text"]').simulate('change', {target: {id: 'name', value: 'test'}});
		wrapper.find('InputText[type="text"]').simulate('change', {target: {id: 'email', value: 'email@email.com'}});
		wrapper.find('InputText[type="text"]').simulate('change', {target: {id: 'password', value: 'password'}});
		wrapper.find('InputText[type="text"]').simulate('change', {target: {id: 'password2', value: 'password'}});
		wrapper.find('form').simulate('submit', {preventDefault() {}});
		expect(wrapper.instance().props.registerUser).toHaveBeenCalled()
	});

	afterEach(() => {
		jest.resetAllMocks();
	});
});