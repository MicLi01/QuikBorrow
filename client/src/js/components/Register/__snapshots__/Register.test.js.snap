// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register renders 1`] = `
<div>
  <a
    href="/"
    onClick={[Function]}
  >
     Back to home
  </a>
  <br />
  <div
    className="sc-bwzfXH gZvuZl"
  >
    <img
      alt="logo"
      height="180"
      src="quikLogo.png"
      width="275"
    />
  </div>
  <div
    className="sc-bdVaJa fKOCmO"
  >
    <h2>
      New User? Register below:
    </h2>
    <p>
      Already have an account? 
      <a
        href="/login"
        onClick={[Function]}
      >
        Log in
      </a>
    </p>
    <form
      noValidate={true}
      onSubmit={[Function]}
    >
      <span
        className="p-float-label"
      >
        <input
          className="p-inputtext p-component"
          error=""
          id="name"
          onChange={[Function]}
          onInput={[Function]}
          onKeyPress={[Function]}
          size="35"
          type="text"
          value=""
        />
        <span>
          
        </span>
        <label
          htmlFor="name"
        >
          Name
        </label>
      </span>
      <br />
      <span
        className="p-float-label"
      >
        <input
          className="p-inputtext p-component"
          error=""
          id="email"
          onChange={[Function]}
          onInput={[Function]}
          onKeyPress={[Function]}
          size="35"
          type="email"
          value=""
        />
        <span>
          
        </span>
        <label
          htmlFor="email"
        >
          Email
        </label>
      </span>
      <br />
      <span
        className="p-float-label"
      >
        <input
          className="p-inputtext p-component"
          error=""
          id="password"
          onChange={[Function]}
          onInput={[Function]}
          onKeyPress={[Function]}
          size="35"
          type="password"
          value=""
        />
        <span>
          
        </span>
        <label
          htmlFor="password"
        >
          Password
        </label>
      </span>
      <br />
      <span
        className="p-float-label"
      >
        <input
          className="p-inputtext p-component"
          error=""
          id="password2"
          onChange={[Function]}
          onInput={[Function]}
          onKeyPress={[Function]}
          size="35"
          type="password"
          value=""
        />
        <label
          htmlFor="password2"
        >
          Confirm Password
        </label>
      </span>
      <br />
      <div>
        <button
          className="p-button p-component p-button-rounded p-button-text-only"
          type="submit"
        >
          <span
            className="p-button-text p-c"
          >
            Sign Up
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;
